lightning-block-sync/src/rest.rs

```diff
@@ -0,0 +1,108 @@
+use crate::{BlockHeaderData, BlockSource, AsyncBlockSourceResult};
+use crate::http::{BinaryResponse, HttpEndpoint, HttpClient, JsonResponse};
+
+use bitcoin::blockdata::block::Block;
+use bitcoin::hash_types::BlockHash;
+use bitcoin::hashes::hex::ToHex;
+
+use std::convert::TryFrom;
+use std::convert::TryInto;
+
+/// A simple REST client for requesting resources using HTTP `GET`.
+pub struct RestClient {
+	endpoint: HttpEndpoint,
+	client: HttpClient,
+}
+
+impl RestClient {
+	/// Creates a new REST client connected to the given endpoint.
+	pub fn new(endpoint: HttpEndpoint) -> std::io::Result<Self> {
+		let client = HttpClient::connect(&endpoint)?;
+		Ok(Self { endpoint, client })
+	}
+
+	/// Requests a resource encoded in `F` format and interpreted as type `T`.
+	async fn request_resource<F, T>(&mut self, resource_path: &str) -> std::io::Result<T>
+	where F: TryFrom<Vec<u8>, Error = std::io::Error> + TryInto<T, Error = std::io::Error> {
+		let host = format!("{}:{}", self.endpoint.host(), self.endpoint.port());
+		let uri = format!("{}/{}", self.endpoint.path().trim_end_matches("/"), resource_path);
+		self.client.get::<F>(&uri, &host).await?.try_into()
+	}
+}
+
+impl BlockSource for RestClient {
+	fn get_header<'a>(&'a mut self, header_hash: &'a BlockHash, _height: Option<u32>) -> AsyncBlockSourceResult<'a, BlockHeaderData> {
+		Box::pin(async move {
+			let resource_path = format!("headers/1/{}.json", header_hash.to_hex());
+			Ok(self.request_resource::<JsonResponse, _>(&resource_path).await?)
+		})
+	}
+
+	fn get_block<'a>(&'a mut self, header_hash: &'a BlockHash) -> AsyncBlockSourceResult<'a, Block> {
+		Box::pin(async move {
+			let resource_path = format!("block/{}.bin", header_hash.to_hex());
+			Ok(self.request_resource::<BinaryResponse, _>(&resource_path).await?)
+		})
+	}
+
+	fn get_best_block<'a>(&'a mut self) -> AsyncBlockSourceResult<'a, (BlockHash, Option<u32>)> {
+		Box::pin(async move {
+			Ok(self.request_resource::<JsonResponse, _>("chaininfo.json").await?)
```

Comment:
With my bitcoind node, I need to add `rest/` as a prefix for REST commands. So this works for me when I do `"rest/chaininfo.json"` in the request, but otherwise returns `not found`.